### PR TITLE
rebuild the navigation

### LINKED ISSUE #124: Main navigation - accessibility
From e-QIP-prototype-truetandem created by [Jay-Neighbours-I72](https://github.com/Jay-Neighbours-I72) : ryanhofdotgov/e-QIP-prototype-truetandem#2927  

Nav items with children are not clear to screen readers.

- [ ] Can we call these out? Make them say "Open to see subsections"

### LINKED ISSUE #326: Consider changing navigation toggle UI from a link to a button
From e-QIP-prototype-truetandem created by [toolness](https://github.com/toolness) : ryanhofdotgov/e-QIP-prototype-truetandem#1393  

The navigation toggle UI is currently implemented as a `<a href="javascript:;;">`.

As I mentioned in https://github.com/truetandem/e-QIP-prototype/pull/1387#discussion_r120868552, I think this might be better as a `<button>`.  From Marcy Sutton's [Links vs. Buttons in Modern Web Applications](https://marcysutton.com/links-vs-buttons-in-modern-web-applications/):

> The starkest difference between a link and a button to me is that a link navigates the user to a new resource, taking them away from the current context (internal links are the only wrinkle here). A button toggles something in the interface, like a video player; or triggers new content in that same context, like a popup menu using `aria-haspopup`.

It might sound pedantic (I always thought it was) but it has UX implications for users of assistive technologies (ATs):

> If a screen reader user calls tech support and gets instructions to “click the button” in your UI that’s really coded as a link, they may have trouble finding it. Also, consider voice interfaces: if you say a command to click a button but it’s really coded as a link, you might have problems, no?

### LINKED ISSUE #472: unable to navigate to nested sections
**What did you do?**

1. Log in
1. Click `Relationships`
1. Click `Marital & relationship status`

**What did you expect to see?**

That section expand.

**What did you see instead?**

* Browser URL turned into `http://localhost:8080/form/identification/javascript:;;`
* Form was blank

**Additional Information**

This is a regression introduced in the past week or so.

### PR DESCRIPTION
**Pull request by afeld**
Fixes #124. Fixes #326. Fixes #472. Part of #475. Builds on #487.

## TODOs

* [x] Expanded properly on navigation
* [x] Error icons
* [x] Expanding and collapsing
* [x] Expanding and collapsing icons
* [x] Numbering
* [x] Sizing/spacing matches

### CODE REVIEW

**Comment by afeld** [comment]
@bpdesigns How much do we care about the user having the ability to collapse the sidebar sections? I ask because it makes the logic a bit confusing (navigate to `/intro` when they click the section to expand it, but not when they collapse it, etc.) If click-to-collapse wouldn't be missed, thoughts about what we should do with the [angle icons](https://fontawesome.com/icons?d=gallery&q=angle-&c=arrows)?

**Comment by bpdesigns** [comment]
@afeld its important for applicants to be able to close the sidebar sections. Some of the later sections have multiple subsections and they get real long real quick.

**Comment by afeld** [comment]
I don't think I explained it very well. Is it important that they can click the section to close it (while remaining in that section), or is it sufficient that the sections close themselves as you navigate? Here's a capture of the behavior currently in this pull request:

screen recording 2018-06-28 at 02 11 pm

When I'm in `Marital` and click `Relationships`, it goes to the `Introduction`, rather than collapsing the `Relationships` section. Does that make sense? Is that behavior ok?

File: src/components/Form/Dropdown/Dropdown.scss [hunk 1]
```diff
@@ -14,9 +14,9 @@
     }
 
     input.usa-input-success {
-      border-top: 1px solid #5b616b;
-      border-left: 1px solid #5b616b;
-      border-right: 1px solid #5b616b;
+      border-top: 1px solid $color-gray;
+      border-left: 1px solid $color-gray;
+      border-right: 1px solid $color-gray;
```
Comment (afeld) [inline src/components/Form/Dropdown/Dropdown.scss]: Made possible by https://github.com/18F/e-QIP-prototype/pull/487/files#diff-a7853f2519441efc78a2e0478f7f0b90R2.

**Comment by ryanhofdotgov** [comment]
Couple of other things I've noticed after quick look:

* For a section review, clicking on a field error, doesn't bring the browser to the correct spot. It jumps there quickly, but quickly jumps back. See screenshot for example. Clicking on provide your full name won't navigate you to the (middle) name field. Note out-of-place navigation menu that appears as well.

screen shot 2018-07-02 at 1 34 58 pm

* Navigating to the "Review and submit" does from "Information about you" doesn't scroll the navigation menu as expected. The "Review and submit" menu stays hidden, only showing the "Psychological health" menu.

menu-hidden

**Comment by afeld** [comment]
Sorry for the bugs that I should have caught! I need to be better about doing my own QA testing.

> it doesn't seem to handle sub-menus properly

👍 You mean indentation-wise, or something beyond that?

**Comment by afeld** [comment]
> For a section review, clicking on a field error, doesn't bring the browser to the correct spot.

Confirmed - on it.

> Navigating to the "Review and submit" does from "Information about you" doesn't scroll the navigation menu as expected. The "Review and submit" menu stays hidden, only showing the "Psychological health" menu.

Can you check if you're seeing the error reported in #447? The fact it doesn't automatically scroll is expected, as noted [above](#issuecomment-401450644), but will consider how best to handle it.

**Comment by ryanhofdotgov** [comment]
No problem – sorry for not being clear – it was the indentation.

**Comment by ryanhofdotgov** [comment]
Also, there is colored "flair" on the left edge of the menus that I think is missing – when things are selected and when there are errors. Here is a screenshot from prod:
screen shot 2018-07-02 at 3 30 16 pm

**Comment by afeld** [comment]
"Flair" re-added. Re: the other issues you reported, the bad news is that they are also happening on `develop` - see linked issues. The good news is that this pull request didn't introduce them 😁

File: src/boot.jsx [hunk 2]
```diff
@@ -1,3 +1,4 @@
+import 'uswds'
```
Comment (afeld) [inline src/boot.jsx]: Needed for the [accordion](https://designsystem.digital.gov/components/accordions/).

**Comment by afeld** [comment]
Ok, this is ready for another review! The expanding/collapsing behavior of a section with subsections has been added back in, without navigating, using the USWDS [accordion](https://designsystem.digital.gov/components/accordions/) component. This should make the accessibility of the page better. @ryanhofdotgov Mind assisting me with some manual QA?

**Comment by ryanhofdotgov** [comment]
👍I'll take a look this afternoon.

**Comment by afeld** [comment]
Follow-up task:

* [ ] Change `... and ...` in the sidebar to ampersands to improve some text wrapping issues

**Comment by ryanhofdotgov** [comment]
After re-checking out this branch, I'm getting this error from the `css_1` when running `make run`:
```
css_1  | [05:51:09] Using gulpfile /usr/src/app/gulpfile.js
css_1  | [05:51:09] Starting 'clean'...
css_1  | [05:51:10] Finished 'clean' after 239 ms
css_1  | [05:51:10] Starting 'copy'...
css_1  | [05:51:10] Starting 'fonts'...
css_1  | [05:51:10] Starting 'images'...
css_1  | [05:51:10] Finished 'copy' after 419 ms
css_1  | [05:51:10] Finished 'fonts' after 542 ms
css_1  | [05:51:10] Finished 'images' after 741 ms
css_1  | [05:51:10] Starting 'build'...
css_1  | [05:51:10] Finished 'build' after 72 μs
css_1  | => changed: /usr/src/app/src/eqip.scss
css_1  | {
css_1  |   "status": 1,
css_1  |   "file": "/usr/src/app/src/components/Navigation/Navigation.scss",
css_1  |   "line": 94,
css_1  |   "column": 20,
css_1  |   "message": "Undefined variable: \"$sidenav-current-border-width\".",
css_1  |   "formatted": "Error: Undefined variable: \"$sidenav-current-border-width\".\n        on line 94 of src/components/Navigation/Navigation.scss\n        from line 48 of src/eqip.scss\n>>       border-left: $sidenav-current-border-width solid transparent;\n\n   -------------------^\n"
css_1  | }
```
Any ideas?

**Comment by afeld** [comment]
@ryanhofdotgov Try `docker-compose rm css && make run`. Might only be available in the newer version of USWDS, and that would force it to use a new container.

File: src/components/Navigation/Navigation.jsx [hunk 3]
```diff
@@ -1,213 +1,18 @@
...
-          sectionClass={sectionClass}
-          sectionNum={displayNum}/>
-      )
-    })
-
     return (
-      <nav className="form-navigation" role="navigation">
-        {nav.filter(x => !!x)}
+      <nav>
+        <SectionList sections={navigation}/>
```
Comment (line47) [inline src/components/Navigation/Navigation.jsx]: do we need to keep the _role_ and _classname_ here?

File: src/components/Navigation/Navigation.scss [hunk 4]
```diff
@@ -102,133 +70,63 @@
...
-
-    &.open .section-content {
-      animation-name: section-openbox;
-      opacity: 1;
-      transition: opacity 300ms ease-out;
-    }
+  // subnavs
+  .usa-sidenav-sub_list {
+    .section-link {
+      border-left: $sidenav-current-border-width solid transparent;
```
Comment (line47) [inline src/components/Navigation/Navigation.scss]: I don't think this value is defined anywhere. I was getting an error when building locally

Comment (afeld) [reply src/components/Navigation/Navigation.scss]: Try https://github.com/18F/e-QIP-prototype/pull/482#issuecomment-403553284.

File: src/components/Navigation/Navigation.scss [hunk 5]
```diff
@@ -102,133 +70,63 @@
...
   }
 
-  .is-valid {
-    .section-number {
-      color: $eapp-green !important;
-    }
+  .section-link {
+    // override USWDS
+    background: none;
+    font-weight: normal;
```
[the remark this answers is not in the document]
Comment (afeld) [reply src/components/Navigation/Navigation.scss]: Can we do that as a follow-up?

File: src/components/Navigation/Navigation.jsx (hunk 3 above)
Comment (afeld) [reply src/components/Navigation/Navigation.jsx]: Ah, good catch on those - didn't realize that the `className` is still used, or that the `role` [should be there as a fallback](https://www.w3.org/WAI/GL/wiki/Using_HTML5_nav_element#Example:The_.3Cnav.3E_element).

**Comment by ryanhofdotgov** [comment]
I'm seeing a few problems... if you like, you can split off items 1 and 2 and address them separately, but I'm hoping item 3 can be addressed before moving forward. Thanks!

1. Horizontal scrollbars in the menu:
rebuild-nav3-scrollbar

2. Various "accoutrements" not matching what I see in prod. See comparison screenshot between prod and rebuild-nav – red numeral 1, red "flair" to left of "Full name", Review is black vs red. 

prod2-example
rebuild-nav2-example

3. Menus like Submit and Print should be disabled, but aren't – the app's state gets very confused if you click them when you're not supposed to:

screen shot 2018-07-10 at 3 06 23 pm

**Comment by afeld** [comment]
Good catches again! Fixed 2 and 3. Can't reproduce 1...what browser?

**Comment by afeld** [comment]
Hmm, not sure. Ok that we do that as a follow-up?

**Comment by ryanhofdotgov** [comment]
Yeah, I think it is time to merge this ticket and work incrementally on the additional important, but not blocking, defects. Getting the red/green indicators things are going well is a major feature of this iteration of eApp (https://github.com/18F/e-QIP-prototype/issues/179 is related).

I'll open up separate tickets for item 1 in https://github.com/18F/e-QIP-prototype/pull/482#issuecomment-404011676, and also some new ones I noted just now:
* Navigation not scrolling, as you scroll main page (e.g., on a review page with lots of errors). I feel like we talked about this recently, but can't find the conversation.
* Submit menu item in navigation menu showing as grey when on Submit page
* Red "flair" still not showing up as expected – should have red bar to left of Review.

(I'll update with issue numbers shortly)

**Comment by afeld** [comment]
Cool, thanks for all the help on this!

**Pull request merged**